update to datreant.core changes

kain88-de opened the pull request:
The tests are running now but there are still a lot of errors.

kain88-de committed:
update dependencies

kain88-de committed:
remove attachements

kain88-de committed:
fixup imports

tests run now

kain88-de committed:
delete UniverseHound

kain88-de committed:
start limb -> metadata conversion

The definitions aren't complete yet. For one I would like the universedef to be
just a datastore now. So some of it's functionality will go into the Sim object.

kain88-de committed:
remove backend class

kain88-de committed:
simplify metadata further

dotsdl commented:
@kain88-de do you want assistance with this? Happy to help if you want to split the work up.

kain88-de committed:
remove universe definition updates from Sim

kain88-de committed:
fix first tests

kain88-de committed:
fix more tests

kain88-de committed:
remove outdated test

kain88-de commented:
@dotsdl I'm done for today. The atomselection metadata still needs to be updated otherwise all tests pass currently.

Of course, the data limbs still have to be added. But getting the tests to pass is pretty straightforward, which is nice.

kain88-de commented on src/mdsynthesis/treants.py:
```diff
@@ -73,55 +69,32 @@ def universe(self):
-            self.universedef._activate()
-            return self._universe
+            self._universe = mda.Universe(*args, **kwargs)
+        return self._universe
```
@dotsdl  I assume you don't like this too much. Right now we reparse the universe everytime we call `sim.universe`. It's because I'm trying to make `UniverseDefinition` a class that only stores information and is therefore decoupled from `Sim`. Before you had this tight coupling between the UniverseDefinition and the Sim object. That allowed neat things like.

```python
s.universedef.topology=other
s.universe # dynamically reload
s.universe # use cached version
```

That is pretty neat, to be honest. I'm not sure how to best reintroduce that behavior while having `Sim` and `UniverseDefinition` decoupled. My current idea would be to calculate a hash from the UniverseDefinition and only reload the universe when that hash has changed. That might take care of your TODO as well.

kain88-de reviewed (commented):
Some comments, if I used the Metadata class correct, would be nice.

kain88-de commented:
> Happy to help if you want to split the work up.

If you want to do the AtomSelection and maybe a hash for Universedefinition (see comment above) that be great. Also have a look if I was deleting to much.

kain88-de committed:
change allowed failures

kain88-de committed:
fix pep8

kain88-de committed:
cache universe in treant

kain88-de committed:
initialize universe on init

kain88-de committed:
python3 compat

kain88-de committed:
atomselection on readonly doesn't throw on access

it will only throw when we want to write

kain88-de committed:
fixup atom selections

kain88-de committed:
reactivate tests for atom selection except one

kain88-de commented on src/mdsynthesis/treants.py:
```diff
@@ -71,57 +71,31 @@ def universe(self):
-        else:
-            self.universedef._activate()
-            return self._universe
+        _args = self.universedef._args
```
I'm using the args as a hash now. This ensures the universe is updated when topology or trajectory change.

dotsdl replied:
I don't think we want this in the `__init__`  method, since it means we will be loading a Universe on init of the `Sim` if it's defined. The point of the `Sim.universe` property is that this loads the universe on access if it's not already loaded, since for big universes or those with lots of trajectories that need to be indexed, this can be very slow.

kain88-de replied:
this is not in the init method. But we can reintroduce the lazy loading. I didn't think about that when I made the change.

dotsdl replied:
@kain88-de because I just removed it. Got lazy loading back so we're good on this front. Working on discover mechanics now.

kain88-de commented on src/mdsynthesis/tests/test_treants.py:
```diff
@@ -353,8 +319,7 @@ def test_fresh_sim_readonly(self, sim):
+    @pytest.mark.skip("impl in datreant.core")
+    def test_write_as_readonly(self, sim):
+        with pytest.raises(IOError):
+            sim.atomselections['foo'] = 'bar'
```
Here I change the existing behavior. In my opinion, just reading from an `AtomSelections` object should be ok even in a read-only SIM. An error should only be thrown when we want to write a new selection. This also means that if the statefile wasn't written we return a default atomselection. I think this behavior is universal to any metadata and would be best implemented in the core library. @dotsdl  what do you think?

dotsdl replied:
@kain88-de I agree with this change. Not exactly sure where to put the behavior you describe in the core library, but being able to handle missing statefiles gracefully and predictably is something we generally prefer.

kain88-de replied:
Do we have tests for readonly in the core library?

dotsdl replied:
@kain88-de we do [here](https://github.com/datreant/datreant.core/blob/develop/src/datreant/core/tests/test_treants.py#L493). I think it's good to leave this test in, so I'm going to remove the skip decorator.

kain88-de replied:
Does this test even work right now?

dotsdl replied:
Ah...not on Python 2.7. This yields an ``OSError`` instead of an ``IOError`` there. Accept either?

kain88-de replied:
yeah it passes. I'm surprised

kain88-de replied:
I think I would like in addition to have a test like this.

```python
    def test_read_as_readonly(self, sim):
        assert sim.atomselections['foo'] is None

```

The current error message isn't so informative. I get a PermissionError as expected. But I'm not 100% set that returning `None` would be better. I also note that this issue is not a blocker for me to merge this.

kain88-de commented:
@dotsdl how do you want to handle the `discover` process? 

```python
sims = mds.discover()
alltreants = dtr.discover()
# Now are sims a subset of alltreants?
```

kain88-de commented:
I think I need to ensure that `universedef.json` is written to mark the directory for good. I think now I can generate a sim with `s = mds.Sim()` and only get a normal Treant until I call `s.universe = u`.

kain88-de commented:
Also the atomselections work now.

kain88-de committed:
turn warnings into errors

kain88-de commented on src/mdsynthesis/treants.py:
```diff
@@ -71,57 +71,31 @@ def universe(self):
+        self._universe = universe
 
     @universe.deleter
     def universe(self):
```
what is the purpose of this delete? Can it be removed?

dotsdl replied:
It makes `del sim.universe` clear the Universe definition, but I'll admit that's a very confusing way to treat `del`. We should remove this.

kain88-de replied:
do you still want to remove this? I would to he honest I don't see the use for `del sim.universe` wouldn't `sim.universe = other_universe` make more sense for a user?

dotsdl replied:
``sim.universe = None`` would be the approach we should encourage for unsetting the universe definition. Using `del` here is just a very confusing usage, outside of what one expects `del` to actually do in Python.

kain88-de commented:
Ping @dotsdl

dotsdl commented:
Hey @kain88-de, this is on my docket for this week. Apologies for the delay.

dotsdl reviewed (changes requested):
Very good stuff here. I think the `Sim.__init__` comment is the only one that requires further work. Thanks for cleaning out the old cruft, @kain88-de!

dotsdl committed:
Removed old skipif for old MDAnalysis in tests

dotsdl commented:
Re [discover](https://github.com/datreant/MDSynthesis/pull/70#issuecomment-345179902).

Good question. Perhaps we should make the JSON files specific to MDS go into an `mdsynthesis` directory inside `.datreant`? We can create this directory on init of `mds.Sim` if it doesn't already exist. Gonna work on this now.

dotsdl committed:
MDSynthesis Sims now create `mdsyntehsis` dir inside `.datreant`

kain88-de commented:
> Good question. Perhaps we should make the JSON files specific to MDS go into an mdsynthesis directory inside .datreant? We can create this directory on init of mds.Sim if it doesn't already exist. Gonna work on this now.

yeah sure we can create a directory.

But the behavior? I'm actually OK with the API I posted. Since a sim is a treant `dtr.discover` should find it. But `mds.discover` should only find sim objects. This means we have to slightly modify discover.

dotsdl commented:
Cool. We're on the same page. Writing an `mds.discover` that uses `dtr.discover`, then filters down to only the Sims. Will need to do something similar with `Bundle`.

dotsdl committed:
We now have mds.Bundle and mds.discover.

Plus a bunch of other changes. Still need to make Bundle properly behave
the same way as dtr.Bundle.

richardjgowers commented:
`mds.discover` only finding mds treants makes a lot of sense.

I'm not sure of the point of `mds.Bundle`?  Plus silently not adding objects is going to lead to some headaches/heartbreaks in the future.  If you end up with a Bundle of mds and dtr Treants then that's the user's problem not ours (maybe they intended it?)

dotsdl committed:
No longer need a separate mds.Bundle.

Minor fixes in datreant.core.Bundle allow for Sims to be handled
gracefully, so long as they are passed in as objects and not paths.
I think this is acceptable because first and foremost, a Sim is a
Treant, so passing in paths only will give you a Bundle of base Treants.

Slicing and filtering on Bundles will always yield Bundles with the
class of the members preserved, since these are held onto in the
Bundle's cache and passed around.

dotsdl committed:
Removed mds.Bundle, added test for mds.discover.

dotsdl commented:
@richardjgowers I like this approach. I rolled with it. Minor changes in `datreant.core` allow for this to work as you describe. `mds.discover` will only ever yield `Bundle`s with `Sim` objects, but these can be used interchangeably with `Treant`s and any other `Bundle`. Thanks for the push!

dotsdl commented:
On the home stretch. Now working on the docs the make this behavior clear. In cutting mode. The next (final?) step will be shoving `datreant.data` into the repository, since limbs are no longer dynamically attachable.

dotsdl committed:
Fixed broken references in the docs.

dotsdl committed:
Small fix for Python 2

dotsdl committed:
Added Python 3 versions to travis

dotsdl commented:
Ready for review. 

Next I'm drafting a blog post for the `datreant` website laying out the changes coming to `datreant`, including the deprecation of `datreant.data` and its merging into `mdsynthesis`. This post will include the motivations for these changes.

After this post is drafted I'll set up a PR for the `datreant.data` merge. This will include a full update to the docs, since `mdsynthesis` will function as much more of a self-contained package built solely on top of `datreant.core`.

Sound good? There's a lot of moving parts here, but trying to hit everything in a reasonable order.

kain88-de commented:
@dotsdl do you know why coverage dropped?

kain88-de commented:
I couldn't spot anything in this pr now. @richardjgowers do you want to have a look?

dotsdl commented:
@kain88-de on coverage, not sure. This PR increases coverage by 0.23% from what I can tell. The `codecov/patch` indicator says that 88.41% of new code added in this PR is covered by tests.

kain88-de reviewed (commented):
Looks good to me. I just have two questions left and once they are answered this is ready to merge.

dotsdl reviewed (commented):
Added some commentary; about to push removal of the skip.

dotsdl committed:
Removed skip of read only test.

kain88-de commented:
> ``sim.universe = None`` would be the approach we should encourage for unsetting the universe definition. Using `del` here is just a very confusing usage, outside of what one expects `del` to actually do in Python.

If this doesn't do what del is supposed to do I think we should remove it. With the new release that has big changes anyway internally it's ok to have one or two user facing changes.

kain88-de commented:
Can’t we make the oserror an ioerror on python 2.7?

> On 17. Dec 2017, at 23:19, David Dotson <notifications@github.com> wrote:
> 
> @dotsdl commented on this pull request.
> 
> In src/mdsynthesis/tests/test_treants.py:
> 
> > @@ -353,8 +319,7 @@ def test_fresh_sim_readonly(self, sim):
>  
>          assert isinstance(s.universe, mda.Universe)
>  
> -        # since we didn't add any atom selections, should raise KeyError
> -        # would be nice if it DIDN'T behave this way, but lazy loading keeps
> -        # Sim init cheaper
> -        with pytest.raises(KeyError):
> -            s.atomselections
> +    @pytest.mark.skip("impl in datreant.core")
> +    def test_write_as_readonly(self, sim):
> +        with pytest.raises(IOError):
> +            sim.atomselections['foo'] = 'bar'
> Ah...not on Python 2.7. This yields an OSError instead of an IOError there. Accept either?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

dotsdl commented:
Working on changes discussed now.

dotsdl committed:
Added test for read only atomselection; removed universe deleter

dotsdl commented:
I've made changes in line with our discussion. I made that test that looks for an ``IOError`` more permissive (also accepts an ``OSError``). I think this is acceptable for the purposes of the test.

dotsdl commented:
I'm fine with a merge if that's the consensus.

kain88-de commented:
merge ahead
